### ISSUE TITLE
Fatal Python error: Cannot recover from stack overflow

### ISSUE THREAD

**Issue by moi90** [scikit-learn-contrib/hdbscan#149]
I tried to cluster 3,974,973 objects (with 32 features). After some days, the calculation was canceled with the following error:

```
Fatal Python error: Cannot recover from stack overflow.

Current thread 0x00007f6d08f61700 (most recent call first):
  File "/data1/mschroeder/miniconda3/envs/pytorch/lib/python3.5/site-packages/hdbscan/plots.py", line 36 in _recurse_leaf_dfs
  File "/data1/mschroeder/miniconda3/envs/pytorch/lib/python3.5/site-packages/hdbscan/plots.py", line 40 in <listcomp>
  File "/data1/mschroeder/miniconda3/envs/pytorch/lib/python3.5/site-packages/hdbscan/plots.py", line [... same line repeated endlessly]
  ...
Aborted
```

The error leads to the following function:
```
def _recurse_leaf_dfs(cluster_tree, current_node):
    children = cluster_tree[cluster_tree['parent'] == current_node]['child']
    if len(children) == 0:
        return [current_node,]
    else:
        return sum([_recurse_leaf_dfs(cluster_tree, child) for child in children], [])
```

It seems that the maximum stack size of my machine (`ulimit -s`: 8192kb) is not large enough for this recursion depth. While a recursive formulation is very elegant in general, an implementation based on an explicit stack (list) might be better in this case.

I ran a quick test and unfortunately my list-based implementation of a DFS leaf extraction is 5 times as slow as the recursive one. Maybe this can be sped up using Cython.

**Comment by moi90** [comment]
My code to compare the speed can be found [here](https://gist.github.com/moi90/4c8a2f9ed3356a3021c9e701d7b50bf0).

**Comment by lmcinnes** [comment]
Thanks, I'll have to look at this. There may be a while to avoid stack depth internally in Cython.

**Comment by moi90** [comment]
Thanks for the hint about `min_cluster_size` and `min_samples`! I'll look into that.

What do you have in mind to "avoid stack depth internally in Cython"?

**Comment by lmcinnes** [comment]
Presuming that the recursive part can be made entirely in Cython then the recursion will ultimately happen in C and not be subject to Python's stack depth constraints (and since C's stack frame are much lighter it is quite reasonable to expect one can go a lot further). I'm not sure if I can get  a recursive version on working in pure Cython though -- the list building, for example, is rather useful.

**Comment by moi90** [comment]
The question is whether an implementation in "pure" Cython is really needed. I agree that lists are useful.

**Comment by lmcinnes** [comment]
I believe I may have a fix in place now. Let me know if it now works for you.

**Comment by moi90** [comment]
Thanks! I will have a look, but it may take a while.

**Comment by moi90** [comment]
Yes, this works for me for 1M objects. Thank you very much!

**Comment by moi90** [comment]
New situation: It works for 1,196,042 with cluster_selection_method="leaf", but not with "eom". This time the error is in prediction.py:

```
Fatal Python error: Cannot recover from stack overflow.
...
File "/data1/mschroeder/miniconda3/envs/pytorch/lib/python3.5/site-packages/hdbscan/prediction.py", line 96 in _recurse_leaf_dfs
...
```

It seems that multiple implementations of `_recurse_leaf_dfs` are spread over the whole codebase.

**Comment by lmcinnes** [comment]
Sadly yes, there is a bunch of duplication. hdbscan grew out and gained a lot of features that were never initially envisaged, and I have never gotten around to the major refactoring that would make all of this sane.

**Comment by lmcinnes** [comment]
It does look like that method is a thin layer over the cython ``recurse_leaf_dfs`` to provide an API that works better in the prediction case.  I'm surprised that overflowed at all -- it shouldn't have as far as I can see from the current code. What version were you using? If it is a fairly current version there may be something odd going on.

**Comment by moi90** [comment]
You are right, in the master branch, the cython function is called. I used version 0.8.11 from pip (which is one year old, admittedly). I will update and try again.

**Comment by moi90** [comment]
Even 0.8.13 does not use the cython function. I will try with master.
